### PR TITLE
Estute/te 2875

### PR DESCRIPTION
**Pull request by estute**
WIP:
This provides a script for generating a feature toggle report for a given deployment (i.e. devstack, production). It requires two data inputs:
* a directory containing sql data for feature toggle tables from each ida
* a directory containing annotation data for each ida

Using this, the feature toggle reporter should be able to create a RST report linking both data sources and displaying the high level state of the toggles configured in the idas, as well as more fine-grained data. It should provide the information specified here: https://open-edx-proposals.readthedocs.io/en/latest/oep-0017-bp-feature-toggles.html#req-12-report

TODO:
* successfully link the annotation data to toggle state data (pulled from tables). I just need to make a call to code_annotations.cli.generate_docs, but since it is a click command, I need to create a shim layer function to be able to call it. Making the PR now.

### CODE REVIEW

File: scripts/feature_toggle_report_generator.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    def __init__(self, name):
+        self.name = name
+        self.toggle_states = {}
+        self.annotation_report_path = None
+
+    def add_toggle_data(self, dump_file_path):
+        """
+        Given the path to a file containing the SQL dump for a
+        feature toggle type in an Ida, parse out the information relevant
```
Comment (jmbowman) [inline scripts/feature_toggle_report_generator.py]: I can see why you'd want to capitalize it this way for the class name to match PEP 8 recommendations, but it really should be IDA (Independently Deployed Application).

File: scripts/README.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+dates. This data is then gathered into report. The feature toggle reporter uses
+this data.
+
+.. code:: bash
+
+    code_annotations static_find_annotations --config_file feature_annotations.yaml
+
+Rename the resulting yml file `<ida_name>_annotation_data.yml`. This step is
+necessary, as the feature toggler report generator will key off the `ida_name`
+in the filename in order to be able to link this data to the SQL data collected
```
Comment (jmbowman) [inline scripts/README.rst]: If it's been captured as a YAML file, "SQL data" doesn't really seem like an appropriate term.  Maybe "toggle state data"?

File: scripts/feature_toggle_report_generator.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        to each toggle and add it to this Ida.
+        """
+        with io.open(dump_file_path, 'r', encoding='utf-8') as dump_file:
+            dump_contents = yaml.safe_load(dump_file.read())
+        for row in dump_contents:
+            toggle_name = row['fields']['name']
+            toggle_type = row['model']
+            toggle_data = row['fields']
+            toggle = ToggleState(toggle_name, toggle_type, toggle_data)
+            if toggle_type in self.toggle_states.keys():
```
Comment (jmbowman) [inline scripts/feature_toggle_report_generator.py]: This logic could be a little simpler if you initialized `self.toggle_states` as a `defaultdict`.

File: scripts/feature_toggle_report_generator.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        readability
+        """
+        null_or_number = lambda n: n if isinstance(n, int) else 0
+        template_data = {}
+        date_template = '%Y-%m-%d %H:%M'
+        if self.toggle_type == 'waffle.switch':
+            template_data['note'] = self.data['note']
+            template_data['creation_date'] = datetime.datetime.strftime(
+                self.data['created'], date_template
+            )
```
Comment (jmbowman) [inline scripts/feature_toggle_report_generator.py]: Should all the date formatting be done by a helper function?  Might simplify the code a little.

File: scripts/feature_toggle_report_generator.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+    @property
+    def data_for_template(self):
+        """
+        Return a dictionary of this Toggle's data for the report, formatted for
+        readability
+        """
+        null_or_number = lambda n: n if isinstance(n, int) else 0
+        template_data = {}
+        date_template = '%Y-%m-%d %H:%M'
```
Comment (jmbowman) [inline scripts/feature_toggle_report_generator.py]: Is it really safe to omit the time zone?  Might be ok if we explicitly normalize it and state somewhere the time zone in use.

File: scripts/feature_toggle_report_generator.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+                '{}-feature-toggle-state.rst'.format(ida_name), 'ida.tpl',
+                variables={'ida': ida_data}
+            )
+
+
+def add_toggle_state_to_idas(idas, dump_file_path):
+    """
+    Given a dictionary of Idas to consider, and the path to a directory
+    containing the SQL dumps for feature toggles in said Idas, read each dump
+    file, parsing and linking it's data into the Ida associated with it.
```
Comment (jmbowman) [inline scripts/feature_toggle_report_generator.py]: nit: "it's" -> "its"

File: templates/flag.tpl [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    * waffle flag name: {{ flag.name }}
+    * waffle flag state (True=On, False=Off): {{ flag.state }}
+    * waffle flag on for everyone: {{ flag.data_for_template['everyone'] }}
+    * waffle flag on for: {{ flag.data_for_template['percent'] }} %
+    * waffle flag on for tests: {{ flag.data_for_template['testing'] }}
+    * waffle flag on for superusers: {{ flag.data_for_template['superusers'] }}
+    * waffle flag on for staff: {{ flag.data_for_template['staff'] }}
+    * waffle flag on for authenticated users: {{ flag.data_for_template['authenticated'] }}
+    * waffle flag on for the following languages:
```
Comment (jmbowman) [inline templates/flag.tpl]: This seems like a lot of repetition of the phrase "waffle flag", maybe ok to omit that given the context?

File: templates/ida.tpl [hunk 8]
```diff
@@ -0,0 +1,33 @@
+Feature toggles configured for **{{ ida.name }}**:
+--------------------------------------------------
```
Comment (jmbowman) [inline templates/ida.tpl]: Careful here, a long IDA name could break the formatting by exceeding the line of dashes.

File: templates/switch.tpl [hunk 9]
```diff
@@ -0,0 +1,7 @@
+.. admonition:: {{ switch.name }}
+
+    * waffle switch name: {{ switch.name }}
+    * waffle switch state (True=On, False=Off): {{ switch.state }}
```
Comment (jmbowman) [inline templates/switch.tpl]: Can probably omit "waffle switch" here too.

File: scripts/README.rst [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+Feature toggles are first defined in the code of an application. We make use of
+the `code_annotations`_ tool, to annotate the definitions of these components
+with information regarding their purpose, default values and planned expiration
+dates. This data is then gathered into report. The feature toggle reporter uses
+this data.
+
+.. code:: bash
+
+    code_annotations static_find_annotations --config_file feature_annotations.yaml
```
[the remark this answers is not in the document]
Comment (estute) [reply scripts/README.rst]: Working on getting this merged in: https://github.com/edx/edx-toggles/pull/3

File: scripts/README.rst [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+Rename the resulting yml file `<ida_name>_annotation_data.yml`. This step is
+necessary, as the feature toggler report generator will key off the `ida_name`
```
[the remark this answers is not in the document]
Comment (estute) [reply scripts/README.rst]: fixed.

File: scripts/README.rst [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+an application's database
+
+Feature Toggle Annotation Data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Feature toggles are first defined in the code of an application. We make use of
+the `code_annotations`_ tool, to annotate the definitions of these components
+with information regarding their purpose, default values and planned expiration
+dates. This data is then gathered into report. The feature toggle reporter uses
+this data.
```
[the remark this answers is not in the document]
Comment (estute) [reply scripts/README.rst]: fixed.

File: scripts/README.rst [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+given deployment. For more information, please see:
+
+https://open-edx-proposals.readthedocs.io/en/latest/oep-0017-bp-feature-toggles.html
+
+Prerequisites:
+--------------
+
+In order to create a feature toggle report for a given deployment, you need
+two types of data: feature toggle annotation data and feature toggle data from
+an application's database
```
[the remark this answers is not in the document]
Comment (estute) [reply scripts/README.rst]: fixed.

File: scripts/README.rst [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+Feature Toggle SQL Data
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The current state of certain feature toggles (i.e. django-waffle, waffle-utils)
+is not captured in the codebase, as it is decoupled from the deployment of
+code. Rather, it must read from the database of an application.
+
+Assuming you have a provisioned `devstack`_, run the following make commands to
+generate the
```
[the remark this answers is not in the document]
Comment (estute) [reply scripts/README.rst]: fixed.

File: scripts/feature_toggle_report_generator.py (hunk 5 above)
Comment (estute) [reply scripts/feature_toggle_report_generator.py]: I did it for readability. This data is all being pulled from a single deployment. If there are tables storing data in different timezone formats in the same deployment, wouldn't that cause other problems? I might be assuming too much here.

File: scripts/feature_toggle_report_generator.py (hunk 6 above)
Comment (estute) [reply scripts/feature_toggle_report_generator.py]: Argh! I will never be able to get this one!

File: templates/flag.tpl (hunk 7 above)
Comment (estute) [reply templates/flag.tpl]: fixed.

File: templates/switch.tpl (hunk 9 above)
Comment (estute) [reply templates/switch.tpl]: fixed.

File: templates/ida.tpl (hunk 8 above)
Comment (estute) [reply templates/ida.tpl]: Is there a way to make the header marker dynamic in length?

File: scripts/feature_toggle_report_generator.py (hunk 5 above)
Comment (jmbowman) [reply scripts/feature_toggle_report_generator.py]: Django typically normalizes to UTC in the database, and converts to the time zone declared in the Django settings when loading datetime objects back from the database.  So they should be consistent by default, but they may not be in the time zone a user would expect.

File: templates/ida.tpl (hunk 8 above)
Comment (jmbowman) [reply templates/ida.tpl]: Something like `{% for character in ida.name %}-{% endfor %}`?

File: scripts/feature_toggle_report_generator.py (hunk 3 above)
Comment (estute) [reply scripts/feature_toggle_report_generator.py]: Good call.

File: scripts/feature_toggle_report_generator.py (hunk 1 above)
Comment (estute) [reply scripts/feature_toggle_report_generator.py]: Yeah, having an acronym in the python tokens has been driving me crazy. Will do.

File: templates/ida.tpl (hunk 8 above)
Comment (estute) [reply templates/ida.tpl]: oh, ok, yeah. I wasn't sure if there was some premade tool for that.

File: scripts/feature_toggle_report_generator.py (hunk 5 above)
Comment (estute) [reply scripts/feature_toggle_report_generator.py]: that's fair. Fixing.

File: templates/ida.tpl (hunk 8 above)
Comment (estute) [reply templates/ida.tpl]: That solution ended up being kind of gross, but as long as this is going to always be viewed as rst rendered to html, it should be ok

File: .gitignore [hunk 15]
```diff
@@ -77,3 +77,7 @@ tests/__init__.py
 
 # Development task artifacts
 default.db
+
+# Used in the creation of feature toggle reports
+annotations/
+reports/
```
[the remark this answers is not in the document]
Comment (estute) [reply .gitignore]: Fixed.

**Pull request merged**